modules/drivers/presto-jdbc/src/metabase/driver/presto_jdbc.clj

```diff
@@ -0,0 +1,470 @@
+(defmethod driver/describe-database :presto-jdbc
+  [driver {{:keys [catalog schema] :as details} :details :as database}]
+  (with-open [conn (-> (sql-jdbc.conn/db->pooled-connection-spec database)
+                       jdbc/get-connection)]
+    (let [schemas (all-schemas driver conn catalog)]
+      {:tables (reduce set/union schemas)})))
+
+(defmethod driver/describe-table :presto-jdbc
+  [driver {{:keys [catalog] :as details} :details :as database} {schema :schema, table-name :name}]
+  (with-open [conn (-> (sql-jdbc.conn/db->pooled-connection-spec database)
+                     jdbc/get-connection)]
+    (let [sql (presto-common/describe-table-sql driver catalog schema table-name)]
+      (log/trace (trs "Running statement in describe-table: {0}" sql))
+      {:schema schema
+       :name   table-name
+       :fields (into
+                 #{}
+                 (map-indexed (fn [idx {:keys [column type] :as col}]
+                                {:name column
+                                 :database-type type
+                                 :base-type         (presto-common/presto-type->base-type type)
+                                 :database-position idx}))
+                 (jdbc/reducible-query {:connection conn} sql))})))
+
+;;; +----------------------------------------------------------------------------------------------------------------+
+;;; |                                            sql-jdbc implementations                                            |
+;;; +----------------------------------------------------------------------------------------------------------------+
+
+(defmethod sql-jdbc.execute/prepared-statement :presto-jdbc
+  [driver ^Connection conn ^String sql params]
+  ;; with Presto JDBC driver, result set holdability must be HOLD_CURSORS_OVER_COMMIT
+  ;; defining this method simply to omit setting the holdability
+  (let [stmt (.prepareStatement conn
+                                sql
+                                ResultSet/TYPE_FORWARD_ONLY
+                                ResultSet/CONCUR_READ_ONLY)]
+       (try
+         (try
+           (.setFetchDirection stmt ResultSet/FETCH_FORWARD)
+           (catch Throwable e
+             (log/debug e (trs "Error setting prepared statement fetch direction to FETCH_FORWARD"))))
+         (sql-jdbc.execute/set-parameters! driver stmt params)
+         stmt
+         (catch Throwable e
+           (.close stmt)
+           (throw e)))))
+
+
+(defmethod sql-jdbc.execute/statement :presto-jdbc
+  [_ ^Connection conn]
+  ;; and similarly for statement (do not set holdability)
+  (let [stmt (.createStatement conn
+                               ResultSet/TYPE_FORWARD_ONLY
+                               ResultSet/CONCUR_READ_ONLY)]
+       (try
+         (try
+           (.setFetchDirection stmt ResultSet/FETCH_FORWARD)
+           (catch Throwable e
+             (log/debug e (trs "Error setting statement fetch direction to FETCH_FORWARD"))))
+         stmt
```

Comment:
I'm not sure this outer `try` is really needed. (I know this was copied this from the old `:presto` driver code, but now that I look at it again I don't think it makes sense, so better to fix past mistakes than to perpetuate them I think.) There's no possible situation where that the body is going to throw an Exception since the only thing we do is `.setFetchDirection` and that's already wrapped in its own try-catch.